File: src/gfx/text/wrap.ts
/**
 * Greedy line breaking over source-offset slices. Lines are [start, end)
 * ranges into the original string. The separators consumed at a break (the
 * spaces and the newline) fall in the gap between consecutive ranges, so
 * selections and copies always map back to exact source offsets.
 *
 * Both terminals implement this identically (go/terminal/text/wrap.go).
 */

export interface WrapLine {
  /** UTF-16 offsets into the source string; display text = s.slice(start, end) */
  start: number;
  end: number;
}

export function wrapText(measure: (s: string) => number, text: string, maxW: number): WrapLine[] {
  const lines: WrapLine[] = [];
  let p = 0;
  for (;;) {
    const nl = text.indexOf('\n', p);
    const pEnd = nl < 0 ? text.length : nl;
    wrapParagraph(measure, text, p, pEnd, maxW, lines);
    if (nl < 0) break;
    p = nl + 1;
  }
  return lines;
}

const isHigh = (cu: number): boolean => cu >= 0xd800 && cu <= 0xdbff;

function wrapParagraph(
  measure: (s: string) => number,
  text: string,
  start: number,
  end: number,
  maxW: number,
  out: WrapLine[],
): void {
  if (start >= end) {
    out.push({ start, end: start }); // empty line (blank paragraph, trailing \n)
    return;
  }
  let i = start;
  while (i < end) {
    // longest prefix that fits - measurement is monotonic, so binary search
    let lo = i + 1;
    let hi = end;
    while (lo < hi) {
      const mid = (lo + hi + 1) >> 1;
      if (measure(text.slice(i, mid)) <= maxW) lo = mid;
      else hi = mid - 1;
    }
    let brk = lo;
    // never split a surrogate pair
    if (brk < end && isHigh(text.charCodeAt(brk - 1))) brk = brk - 1 > i ? brk - 1 : brk + 1;
    if (brk < end) {
      // back off to the last break opportunity (either side of a space)
      // so words stay whole
      // if none then hard break mid-word
      let sp = -1;
      for (let j = brk; j > i; j--) {
        if (text.charCodeAt(j - 1) === 32 || text.charCodeAt(j) === 32) {
          sp = j;
          break;
        }
      }
      if (sp > i) brk = sp;
    }
    // the line ends before its trailing spaces
    // the next starts after them
    let lineEnd = brk;
    while (lineEnd > i && text.charCodeAt(lineEnd - 1) === 32) lineEnd--;
    out.push({ start: i, end: lineEnd > i ? lineEnd : brk });
    i = brk;
    while (i < end && text.charCodeAt(i) === 32) i++;
  }
}

File: go/terminal/text/wrap.go
package text

import (
	"strings"
	"unicode/utf8"
)

// Line is one wrapped line as a [Start, End) byte range into the source
// string. The separators consumed at a break (the spaces and the newline)
// fall in the gap between consecutive ranges, so selections and copies map
// back to exact source offsets. Mirrors src/gfx/text/wrap.ts.
type Line struct{ Start, End int }

// Wrap greedily breaks s to fit maxW, honoring '\n'. measure must be
// monotonic in prefix length (it is: it's an advance sum).
func Wrap(measure func(string) float32, s string, maxW float32) []Line {
	var out []Line
	p := 0
	for {
		nl := strings.IndexByte(s[p:], '\n')
		pe := len(s)
		if nl >= 0 {
			pe = p + nl
		}
		wrapPara(measure, s, p, pe, maxW, &out)
		if nl < 0 {
			break
		}
		p = pe + 1
	}
	return out
}

func wrapPara(measure func(string) float32, s string, start, end int, maxW float32, out *[]Line) {
	if start >= end {
		*out = append(*out, Line{start, start}) // empty line (blank paragraph, trailing \n)
		return
	}
	// Rune boundaries as byte offsets: bounds[i] is where rune i starts;
	// bounds[n] == end.
	var bounds []int
	for i := start; i < end; {
		bounds = append(bounds, i)
		_, sz := utf8.DecodeRuneInString(s[i:end])
		i += sz
	}
	bounds = append(bounds, end)
	n := len(bounds) - 1

	i := 0
	for i < n {
		// Longest prefix that fits - binary search over rune count.
		lo, hi := i+1, n
		for lo < hi {
			mid := (lo + hi + 1) >> 1
			if measure(s[bounds[i]:bounds[mid]]) <= maxW {
				lo = mid
			} else {
				hi = mid - 1
			}
		}
		brk := lo
		if brk < n {
			// Back off to the last break opportunity - either side of a space
			// so words stay whole. None means a hard break mid-word.
			sp := -1
			for j := brk; j > i; j-- {
				if s[bounds[j]-1] == ' ' || s[bounds[j]] == ' ' {
					sp = j
					break
				}
			}
			if sp > i {
				brk = sp
			}
		}
		// The line ends before its trailing spaces; the next starts after them.
		lineEnd := brk
		for lineEnd > i && s[bounds[lineEnd]-1] == ' ' {
			lineEnd--
		}
		le := bounds[lineEnd]
		if lineEnd == i {
			le = bounds[brk]
		}
		*out = append(*out, Line{bounds[i], le})
		i = brk
		for i < n && s[bounds[i]] == ' ' {
			i++
		}
	}
}
